components/ca-select.js

```diff
@@ -0,0 +1,367 @@
+                                };
+                            }
+                        }
+                    }
+                }
+                
+                // just in case this is a re-render, clear this!
+                this.removeAttribute('data-empty');
+            } else {
+
+                // add an attribute to allow empty selects know this item has no children
+                this.setAttribute('data-empty', 'true');
+            }
+        };
+
+        /**
+         * Executes each time an option is clicked
+         * @returns {void} void
+         */
+        checkChange(el) {
+
+            if (!el || el.value === '') return;
+
+            const selected = (el.getAttribute('data-selected') === 'true');
+            el.setAttribute("data-selected", !selected);
+        };
+
+        /**
+         * A seemingly pointless method
+         * @returns {void} void
+         */
+        onupdate(data) {
+
+            helpers.trace(data);
+        };
+
+        /**
+         * onChange handler
+         * @returns {void} void
+         */
+        onChangeHandler() {
+
+            const parentForm = parentByAttribute(this, 'tagName', 'CA-FORM');
+
+            if (parentForm) {
+
+                parentForm.setAttribute('data-select-' + this.id, this.value);
+            }
+        }
+
+        get data() {
+
+            return this._data || [];
+        }
+
+        set data(value) {
+
+            this._data = value || [];
+            this.render();
+        }
+
+        set schema(value) {
+
+            let selectData = [];
+            const data = schema.enum || schema.type;
+
+            if (Array.isArray(data) && data.length) {
+
+                for (let i = 0, l = data.length; i < l; i++) {
+
+                    const value = data[i].enum && data[i].enum[0] || data[i];
+                    const label = data[i].title || (l == 1 && schema.title) || data[i];
+                    let links = data[i].links || (l == 1 && schema.links) || [];
+                    links = links.filter(link => link.rel === 'preview');
+
+                    selectData.push({
+                        value: value,
+                        label: label,
+                        attachment: links.length && links[0] || null
+                    });
+                }
+            }
+
+            this.data = selectData;
+        }
+
+        get src() {
+
+            return this.getAttribute('schema-url');
+        }
+
+        set src(value) {
+
+            this.setAttribute('schema-url', value);
+            this.loadSchema();
+        }
+
+        get value() {
+
+            if (this.multiple) {
+
+                let results = [];
+
+                if (this.root) {
+
+                    const options = (!isTouch) ? this.root.querySelectorAll('[data-selected="true"]') : Array.prototype.slice.call(this.root.options).where('selected', true);
+
+                    helpers.forEach(options, item => {
```

Comment:
helpers.forEach should be put into a new for-each module and added as a dependency